include/boost/sam/basic_shared_lock.hpp

```diff
@@ -0,0 +1,206 @@
+struct basic_shared_lock
+{
+  using executor_type = net::any_io_executor;
+  
+  /// Construct an empty lock_guard.
+  basic_shared_lock()                   = default;
+  basic_shared_lock(const basic_shared_lock &) = delete;
+  basic_shared_lock(basic_shared_lock &&lhs)
+      : mtx_(lhs.mtx_), executor_(std::move(lhs.executor_)), locked_(lhs.locked_)
+  {
+    lhs.mtx_ = nullptr;
+    lhs.locked_ = false;
+  }
+
+  template <typename Executor_>
+  basic_shared_lock(basic_shared_lock<Executor_> &&lhs,
+                    typename std::enable_if<std::is_convertible<Executor_, executor_type>::value>::type * = nullptr)
+      : mtx_(lhs.mtx_), executor_(std::move(lhs.executor_)), locked_(lhs.locked_)
+  {
+    lhs.mtx_ = nullptr;
+    lhs.locked_ = false;
+  }
+
+  basic_shared_lock &operator=(const basic_shared_lock &) = delete;
+  basic_shared_lock &operator=(basic_shared_lock &&lhs)
+  {
+    std::swap(lhs.mtx_, mtx_);
+    std::swap(lhs.locked_, locked_);
+    std::swap(lhs.executor_, executor_);
+    return *this;
+  }
+
+  template <typename Executor_>
+  auto operator=(basic_shared_lock<Executor_> &&lhs)
+      -> typename std::enable_if<std::is_convertible<Executor_, executor_type>::value, basic_shared_lock>::type &
+  {
+    std::swap(lhs.mtx_, mtx_);
+    std::swap(lhs.locked_, locked_);
+    std::swap(lhs.executor_, executor_);
+    return *this;
+  }
+
+  /// Unlock the underlying mutex.
+  ~basic_shared_lock()
+  {
+    if (mtx_ != nullptr && locked_)
+      mtx_->unlock_shared();
+  }
+
+  template <typename Executor_>
+  basic_shared_lock(basic_shared_mutex<Executor_> &mtx,
+                    typename std::enable_if<std::is_convertible<Executor_, executor_type>::value>::type * = nullptr)
+      : mtx_(&mtx.impl_), executor_(mtx.get_executor())
+  {
+    mtx.lock_shared();
+    locked_ = true;
+  }
+
+  template <typename Executor_>
+  basic_shared_lock(basic_shared_mutex<Executor_> &mtx,
+                    typename std::enable_if<
+                        std::is_convertible<Executor_, executor_type>::value,
+                        std::defer_lock_t>::type) noexcept
+      : mtx_(&mtx.impl_), executor_(mtx.get_executor())
+  {
+  }
+
+  template <typename Executor_>
+  basic_shared_lock(basic_shared_mutex<Executor_> &mtx,
+                    typename std::enable_if<
+                        std::is_convertible<Executor_, executor_type>::value,
+                        std::adopt_lock_t>::type) noexcept
+      : mtx_(&mtx.impl_), executor_(mtx.get_executor()), locked_(true)
+  {
+  }
+
+  template <typename Executor_>
+  basic_shared_lock(basic_shared_mutex<Executor_> &mtx,
+                    typename std::enable_if<
+                        std::is_convertible<Executor_, executor_type>::value,
+                        std::try_to_lock_t>::type) noexcept
+      : mtx_(&mtx.impl_), executor_(mtx.get_executor()), locked_(mtx.try_lock_shared())
+  {
+  }
+
+  void unlock(error_code & ec)
```

Comment:
Does having error_code / throwing variants for an unlock operation make any sense? Unlocking is the kind of function you'd expect to be able to call from a destructor, without throwing. The two throwing cases here could be stated as preconditions.

Reply:
it does because you can call it on a lock that doesn't own a mutex or didn't lock that mutex. Then it throws, just like std::unique_lock